lib/internal/fs/promises.js

```diff
@@ -696,27 +697,54 @@ async function mkdir(path, options) {
 
 const opendir = promisify(_opendir);
 
+async function readdirRecursive (origPath, options) {
+  async function _r (path) {
+    const result = await binding.readdir(
+      pathModule.toNamespacedPath(path),
+      options.encoding,
+      !!options.withFileTypes,
+      kUsePromises
+    );
+
+    return options.withFileTypes ?
+      Promise.all(
+        (await getDirectoryEntriesPromise(path, result)).map(
+          async (dirent) => ([
+            dirent,
+            ...( dirent.isDirectory() ? (await _r(pathModule.join(path, dirent.name), options)) : []).flat()
+          ]))
+        )
+     : Promise.all(
+      result.map(async (ent) => {
+        const innerPath = pathModule.join(path, ent)
+        const relativePath = pathModule.relative(origPath, innerPath)
+        const stat = await lstat(innerPath);
+        return [
+          relativePath,
+          ...( stat.isDirectory() ? (await _r(innerPath, options)).flat() : [])
+        ]
+      })
+    )
+  }
```

Comment:
I fear that this will result in a recursive promise chain that could be very costly and poorly performing. I do not have quantitative feedback right now, so I'm not blocking this. From experience, this kind of code performs poorly, and it's likely to generate `EBUSY` errors.

I would have implemented this PR as a linear algorithm, but I think we should land this PR asap and iterate on it later if performance becomes an issue.

Reply:
simplified. let me know what you think

Reply:
This is still recursive. It should be possible to write this as a loop instead. I'm ok with the current implementation, but please add a TODO to make it iterative.

Reply:
adding a comment for this now.